sdk/python/kfp/compiler/compiler_utils.py

```diff
@@ -325,9 +326,191 @@ def get_inputs_for_all_groups(
     return inputs
 
 
+class InvalidTopologyException(Exception):
+    pass
+
+
+def validate_parallel_for_fan_in_consumption_legal(
+    consumer_task_name: str,
+    upstream_groups: List[str],
+    group_name_to_group: Dict[str, tasks_group.TasksGroup],
+) -> None:
+    """Checks that a dsl.Collected object is being used results in an
+    unambiguous pipeline topology and is therefore legal.
+
+    Args:
+        consumer_task_name: The name of the consumer task.
+        upstream_groups: The names of the producer task's upstream groups, ordered from outermost group at beginning to producer task at end. This is produced by produced by _get_uncommon_ancestors.
+        group_name_to_group: Map of group name to TasksGroup, for fast lookups.
+    """
+    # illegal if the producer has a parent conditional outside of its outermost for loop, since the for loop may or may or may not be executed
+    # for example, what happens if text == 'b'? the resulting execution behavior is ambiguous.
+    #
+    # @dsl.pipeline
+    # def my_pipeline(text: str = ''):
+    #     with dsl.Condition(text == 'a'):
+    #         with dsl.ParallelFor([1, 2, 3]) as f:
+    #             t = double(num=f)
+    #     x = add(nums=dsl.Collected(t.output))
+
+    for group_name in upstream_groups:
+        group = group_name_to_group[group_name]
+        if group.group_type == tasks_group.TasksGroupType.FOR_LOOP:
+            break
+        if group.group_type in [
+                tasks_group.TasksGroupType.CONDITION,
+                tasks_group.TasksGroupType.EXIT_HANDLER,
+        ]:
+            raise InvalidTopologyException(
+                f'{ILLEGAL_CROSS_DAG_ERROR_PREFIX} When using dsl.{for_loop.Collected.__name__} to fan-in outputs from a task within a dsl.{tasks_group.ParallelFor.__name__} context, the dsl.{tasks_group.ParallelFor.__name__} context manager cannot be nested within a dsl.{group.__class__.__name__} context manager unless the consumer task is too. Task {consumer_task_name} consumes from {upstream_groups[-1]} within a dsl.{group.__class__.__name__} context.'
+            )
```

Comment:
IIUC, there's no other type of group we could possible meet here, right? In that case, we should be explicit on raising unexpected type error for the final else. 
That also means we would never need to look at `upstream_groups[1]` and beyond. Should we be explicit and pass in only what we need, which is always `upstream_groups[0]` or null?

Reply:
Yes, that's a good catch. I added some additional validation to this function that warrants keeping the function signature and provided inputs as-is, but I've modified the logic to better reflect the cases we're looking for to align with what you're describing.